modules/reactor/src/meshgenerators/ExtraElementIDConnectionFixup.C

```diff
@@ -0,0 +1,231 @@
+        // with the current coarse element ID
+        // FIXME: checking the connection through line segments in 3D can avoid some unnecessary
+        // assignment.
+        for (const auto & id : int_nodes)
+          for (const auto & elem_id : map_node_to_fine[id])
+          {
+            Elem * elem = mesh->elem_ptr(elem_id);
+            auto old_id = elem->get_extra_integer(src_id);
+            if (old_id != coarse_id)
+            {
+              if (_verbose)
+                _console << " Element with ID " << elem->id() << ": " << old_id << " => "
+                         << coarse_id << std::endl;
+              elem->set_extra_integer(src_id, coarse_id);
+              map_coarse_to_fine[old_id].erase(elem->id());
+              map_coarse_to_fine[coarse_id].insert(elem->id());
+              for (unsigned int n = 0; n < elem->n_nodes(); ++n)
+                visited_nodes.insert(elem->node_ptr(n)->id());
+
+              // put changed ids into re-consideration
+              coarse_ids.insert(old_id);
+            }
+          }
+      }
+      for (auto & id : nodes)
+        visited_nodes.insert(id);
+    }
+  }
+
+  std::map<dof_id_type, Real> element_volume;
+  std::map<dof_id_type, Point> element_centroid;
+  for (auto & elem : mesh->element_ptr_range())
+  {
+    dof_id_type id = elem->get_extra_integer(src_id);
+    if (id == DofObject::invalid_id)
+      continue;
+
+    element_volume[id] += elem->volume();
+    element_centroid[id] += elem->true_centroid() * elem->volume();
+  }
+  for (auto & pair : element_centroid)
+    pair.second /= element_volume[pair.first];
+  if (_verbose)
+    _console << " Number of coarse elements after fixup: " << element_volume.size() << std::endl;
+
+  if (getParam<bool>("order_ids_with_centroid"))
+  {
+    std::vector<std::pair<dof_id_type, Point>> sorted_centroids;
+    sorted_centroids.reserve(element_centroid.size());
+    for (const auto & it : element_centroid)
+      sorted_centroids.push_back(std::make_pair(it.first, it.second));
+    // Our sorting functor
+    auto sort_fun = [](const std::pair<dof_id_type, Point> & lhs,
+                       const std::pair<dof_id_type, Point> & rhs) -> bool {
+      return lhs.second < rhs.second;
+    };
+    std::sort(sorted_centroids.begin(), sorted_centroids.end(), sort_fun);
+    std::map<dof_id_type, dof_id_type> new_element_ids;
+    for (dof_id_type i = 0; i < sorted_centroids.size(); ++i)
+      new_element_ids[sorted_centroids[i].first] = i;
+
+    for (auto & elem : mesh->element_ptr_range())
+    {
+      dof_id_type id = elem->get_extra_integer(src_id);
+      if (id != DofObject::invalid_id)
+        elem->set_extra_integer(src_id, new_element_ids[id]);
+    }
+    if (_verbose)
+      _console << " Reassign IDs from 0 to " << new_element_ids.size() - 1
+               << " to elements based on the sequence of the ordered IDs" << std::endl;
```

Comment:
this feels like a different mesh generator tbh.

Reply:
some sort of ExtraElementIDReorderingGenerator
and this particular one is a centroid comparison

Reply:
Yes. But I feel when others need this capability, we can factor out this easily.

Reply:
This is actually quite a painful decision. We do want a flexible system that users can do lots of things freely with the input, but on the other hand, we do not want the input to be long. I think here we try to have small number of generators and go flexible when needed.

Reply:
I think we want to split this out into another generator. 
OR
rename the whole generator, split the documentation in 3 parts with pictures and say it s a mesh generator that does a bunch of unrelated mesh operations on extra Element IDs.

the first one is much more MOOSE like.